src/qibolab/instruments/qblox/sequencer.py

```diff
@@ -74,6 +76,7 @@ def add_waveforms(
                 waveform_i not in self.unique_waveforms
                 or waveform_q not in self.unique_waveforms
             ):
+                #
```

Comment:
The memory could be computed after the deduplication, no need to do the two things at the same time pulse by pulse (it is also fine, the same amount of operations will be performed, but if we split the two we could keep them more isolated and modular).

The error raised is used only to advance the sequencer used, and requires repeated "attempts" to assign a given pulse to a sequencer.
Also this strategy is not bad, but we could even just eagerly batch the pulses once deduplicated, and assign one sequencer to each batch (again, for isolation and modularity, since batching does not require the sequencers, but only the information on the memory available on each one).